Fix authentication provider exception not logged

BioinformaticsPlatformWIV-ISP opened the pull request:
This a copy of the PR  #7139 closed after a failing rebase tentative.
During an external provider authentication, log and reraise the exception properly instead of ignoring it when the provider raise one.

Commit:
Fix the not logged exeption during AD authentication, now log and reraise it properly

nsoranzo commented:
@BioinformaticsPlatformWIV-ISP I've rebased your PR.

nsoranzo commented:
@galaxybot test this

BioinformaticsPlatformWIV-ISP commented:
@nsoranzo  Thank you for your help,  doing the rebase ourselves was a bad experience.

nsoranzo commented on lib/galaxy/auth/__init__.py:
```diff
@@ -64,8 +64,9 @@ def check_auto_registration(self, trans, login, password, no_password_check=Fals
-                        break
+                    except Conflict as conflict:
+                        log.exception(conflict)
+                        raise
```
If the `raise` here is meant to propagate the error to the user, you need to catch this exception at https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/controllers/user.py#L85

BioinformaticsPlatformWIV-ISP replied:
The exception is now propagated to the user, i.e. converted to a message return by the autoregistration action inside of the user controller.

jvanbraekel committed:
Handle the auto registration conflict error

Catch and display to the user (as message) any conflict error raised by
the authentication providers in case of autoregistration.

nsoranzo commented on lib/galaxy/webapps/galaxy/controllers/user.py:
```diff
@@ -82,33 +82,35 @@ def __autoregistration(self, trans, login, password):
-                        auto_create_groups=autoreg["auto_create_groups"],
-                        auto_create_roles=autoreg["auto_create_roles"],
+        if autoreg :
+            if autoreg["auto_reg"]:
```
You can combine the 2 ifs and save one indentation level:
```python
        if autoreg and autoreg["auto_reg"]:
```

jvanbraekel replied:
I thought about it . However, its imply that the next else have to be updated to an  `elif autoreg: `  (to be able to discriminate the case when autoreg exist and autoreg["auto_reg"] not).  That make this piece of code less intuitive because at the first reading,  its look like the 'same' check is done twice.
That being said, I can go for the 'elif cascade' if you think that indentation level save its wroth the change.

nsoranzo replied:
I don't see an `else` at level of `if autoreg:` (and I don't think one is needed!), so I'm not sure I understand your comment.

jvanbraekel replied:
My comment concern the `else: message = "No such user or invalid password."` If I go for a ` if autoreg and autoreg["auto_reg"]:` like you suggest this one cannot  be place at the same level as the 'if-and' without changes, otherwise `message` would be overridden in case of an conflict exception. It neither can be placed underneath the 'if-and' because it will never fire-up.

nsoranzo replied:
I see, thanks for explaining. I'd suggest an early return then, see suggestion above.

nsoranzo commented on lib/galaxy/webapps/galaxy/controllers/user.py:
```diff
@@ -82,33 +82,35 @@ def __autoregistration(self, trans, login, password):
+            autoreg = trans.app.auth_manager.check_auto_registration(trans, login, password)
+        except Conflict as conflict:
+            message = "Auto-registration failed, {}".format(conflict)
+            autoreg = None
```
```suggestion
            return message, None
```

mvdbeek commented:
Thanks @BioinformaticsPlatformWIV-ISP!
